sdk/storage/azure-storage-blob-batch/src/main/java/com/azure/storage/blob/batch/BlobBatch.java

```diff
@@ -0,0 +1,427 @@
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This class allows for batching of multiple Azure Storage operations in a single request via {@link
+ * BlobBatchClient#submitBatch(BlobBatch)} or {@link BlobBatchAsyncClient#submitBatch(BlobBatch)}.
+ *
+ * <p>Azure Storage Blob batches are homogeneous which means a {@link #deleteBlob(String) delete} and {@link
+ * #setBlobAccessTier(String, AccessTier) set tier} are not allowed to be in the same batch.</p>
+ *
+ * {@codesnippet com.azure.storage.blob.batch.BlobBatch.illegalBatchOperation}
+ *
+ * <p>Please refer to the <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/blob-batch">Azure Docs</a>
+ * for more information.</p>
+ */
+public final class BlobBatch {
+    private static final String X_MS_VERSION = "x-ms-version";
+    private static final String BATCH_REQUEST_CONTENT_ID = "Batch-Request-Content-Id";
+    private static final String BATCH_REQUEST_URL_PATH = "Batch-Request-Url-Path";
+    private static final String CONTENT_ID = "Content-Id";
+    private static final String BATCH_BOUNDARY_TEMPLATE = "batch_%s";
+    private static final String REQUEST_CONTENT_TYPE_TEMPLATE = "multipart/mixed; boundary=%s";
+    private static final String BATCH_OPERATION_CONTENT_TYPE = "Content-Type: application/http";
+    private static final String BATCH_OPERATION_CONTENT_TRANSFER_ENCODING = "Content-Transfer-Encoding: binary";
+    private static final String BATCH_OPERATION_CONTENT_ID_TEMPLATE = "Content-ID: %d";
+    private static final String HTTP_VERSION = "HTTP/1.1";
+    private static final String OPERATION_TEMPLATE = "%s %s %s";
+    private static final String HEADER_TEMPLATE = "%s: %s";
+
+    /*
+     * Track the status codes expected for the batching operations here as the batch body does not get parsed in
+     * Azure Core where this information is maintained.
+     */
+    private static final int[] EXPECTED_DELETE_STATUS_CODES = {202};
+    private static final int[] EXPECTED_SET_TIER_STATUS_CODES = {200, 202};
+
+    private final ClientLogger logger = new ClientLogger(BlobBatch.class);
+
+    private final BlobAsyncClient blobAsyncClient;
+
+    private final Deque<Mono<? extends Response<?>>> batchOperationQueue;
+    private final List<ByteBuffer> batchRequest;
+    private final Map<Integer, BlobBatchOperationResponse<?>> batchMapping;
+
+    private final AtomicInteger contentId;
+    private final String batchBoundary;
+    private final String contentType;
+
+    private BlobBatchType batchType;
+
+    BlobBatch(String accountUrl, HttpPipeline pipeline) {
+        this.contentId = new AtomicInteger(0);
```

Comment:
new AtomicInteger() is sufficient. The default is 0.

Reply:
Will update in Preview 5

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/5756